examples/xtt_asio_server.cpp

```diff
@@ -0,0 +1,318 @@
+    cert_map[xtt::suite_spec::X25519_LRSW_ED25519_AES256GCM_BLAKE2B] = 
+        std::make_unique<xtt::server_certificate_context_ed25519>(serialized_cert, serialized_privkey);
+
+    return 0;
+}
+
+id_provisioner::id_provisioner(boost::asio::ip::tcp::socket tcp_socket,
+                               const std::unordered_map<xtt::suite_spec, std::unique_ptr<xtt::server_certificate_context>>& cert_map,
+                               xtt::server_cookie_context& cookie_ctx,
+                               std::unordered_map<xtt::group_identity, std::unique_ptr<xtt::group_public_key_context>>& gpk_map)
+    : handshake_(std::move(tcp_socket), cert_map, cookie_ctx),
+      gpk_map_(gpk_map),
+      claimed_gid_(),
+      rand_dev_(),
+      unsafe_rng_(rand_dev_())
+{
+}
+
+void id_provisioner::run()
+{
+    auto self = shared_from_this();
+
+    handshake_.async_negotiate([this, self](auto rc,
+                                            auto claimed_gid,
+                                            auto requested_client_id,
+                                            auto closure)
+                               {
+                                   first_callback(rc, claimed_gid, requested_client_id, closure);
+                               });
+}
+
+int
+id_provisioner::assign_client_id(xtt::identity* assigned_client_id_out,
+                                 const xtt::identity& requested_client_id,
+                                 const xtt::group_identity& gid,
+                                 std::unique_ptr<xtt::pseudonym> clients_pseudonym)
+{
+    // In principle, we could use the gid and pseudonym when selecting an id for the client
+    // (i.e., use the gid to choose the pool of id's,
+    // and use the pseudonym to ensure the same client always gets the same id).
+    (void)gid;
+    (void)clients_pseudonym;
+
+    // If the client sent xtt_null_client_id assign them a randomly-generated id.
+    // Otherwise, just echo back what they requested.
+    if (requested_client_id.is_null()) {
+        std::vector<unsigned char> new_id_serialized;
+        std::generate_n(std::back_inserter(new_id_serialized), sizeof(xtt_identity_type), unsafe_rng_);
+        *assigned_client_id_out = xtt::identity(new_id_serialized);
+    } else {
+        *assigned_client_id_out = requested_client_id;
+    }
+
+    return 0;
+}
+
+void id_provisioner::first_callback(xtt::return_code rc,
+                                    const xtt::group_identity& claimed_gid,
+                                    const xtt::identity& requested_client_id,
+                                    xtt::asio::server_handshake* closure)
+{
+    if (xtt::return_code::WANT_VERIFYGROUPSIGNATURE != rc) {
+        std::cerr << "Error with first stage of handshake\n";
+        // self now goes out of scope, releasing its memory (thus closing socket in handshake_)
+        return;
+    } else {
```

Comment:
No `else`.  We've already returned if the `if` succeeded.  Removing the `else` saves one level of indentation and make it easier to stack multiple error checks.